Improving STL compatibility using template meta programming preview #20891

HattrickGenerator opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test
- [x]  sample code can be built with the project CMake

This is how the header only part of forwarding the STL could look like. There are still most algorithms missing, but the idea is there:
Create one specialization for forwarding only the pointers to the opencv iterators and one, for the original input parameters.
Decide if it's valid to forward the raw pointers based on the runtime value of matrices involved (isContinuous flag is on).

A special case is involved for algorithms returning an opencv iterator. Here we still use pointers when possible for maximal efficiency, but we later add the offset to the input iterator for completely abstracting away the internals of our machinery.

This PR serves as an example for how it could be done. If there is an interest I will add the rest of the algorithms, and cases for reverse iterators.


```
force_builders=linux,docs,win64
```

HattrickGenerator committed:
Adding template meta programming backend and a header for forwarding the STL and using it (efficiently!) with openCV iterators

The idea behind it is two fold:
Because we can't know at compile time if our matrix (described by the iterator) is contiguous or not, a function cv::detail::__iterators__replaceable takes a variadic template pack and looks if there are opencv iterators. If yes it checks if they are contiguous.
Only if all present iterators point to a contiguous matrix, we get true. For everything else (including no opencv iterators in the pack) we get false.
This is tested with unit tests in the CORE_stl_forward suite

The second part consists of making a tuple with every iterator replaced by its (casted) pointer.
This part is limited to up to 10 input arguments, as I couldn't make it completely generic due to limitations on decltype(auto) in C++11. Starting from C++14 it is easy to replace however in a completely generic manner.

We then forward either the tuple or the original arguments to the stl function. Like this we get good performance (i.e. pointers) for where its possible, while staying safe with the iterators for all the cases, where this is not possible, combining the best of the two worlds.

Tests for this are added.

A new folder stl can now be found in core, containing the forwarded algorithms. This was changed in the CMakeLists.txt. The utility functions are added to the detail headers in core.

HattrickGenerator committed:
Adding find function as a prototype for stl functions returning iterators.

Here it is a bit more complicated, as we would return a pointer normally. So we differentiate as to when we need to return an opencv iterator by actually letting the algorithm run with pointers and then adding the difference to the begin pointer to the begin iterator, so that from the outside it's not noticeable that the algo runs with pointers instead of iterators.
This requires another constexpr function __get_first_cv_it_index for which tests are added.

alalek commented on modules/core/test/test_stl_forward.cpp:
```diff
@@ -0,0 +1,217 @@
+    EXPECT_FALSE(cv::experimental::__iterators__replaceable(intVec.begin(), intVec.end(), subMat.begin()));
+}
+
+TEST_F(CORE_stl_forward, DISABLED_iterators_replacable_reverse_iterator)
```
Why is this test disabled?

HattrickGenerator replied:
Hello @alalek,
thanks for replying. 
The branch is currently only experimental and there is no support for reverse iterators, yet.
I wanted to test the waters and see, if such a feature would be interesting for the community. If this is the case I am willing to add the missing algorithms (that exist in C++11), add support for the reverse iterators and write some documentation for it.

HattrickGenerator commented:
The tests are still disabled but I added a static_assert for now, that evaluates to false if any of the forwarding is done with reverse iterators (that are not in the feature, yet).

HattrickGenerator commented:
The feature is hard to implement in C++11 only. 
Is there a way to enable it only when a C++14 compiler is available?

HattrickGenerator committed:
Change stl algorithms to use feature testing and only work with C++14 return type deduction auto.
This greatly simplifies the implementation of a large number of internal machinery

HattrickGenerator committed:
Rename the delegated functions with _ and _impl pre- and postfix
Remove the constraint for find to have a default constructable iterator!

HattrickGenerator committed:
Fix the function for finding the first opencv (or reverse opencv!) iterator.
Continue implementing support for reverse iterators

HattrickGenerator committed:
adapting the iterator__replacable function to take reverse iterators into account

HattrickGenerator committed:
Adding proper reverse iterator support also for find algorithm (that returns an iterator)

HattrickGenerator commented:
I re-worked my commit. It proved to time consuming to implement in C++11. So I used the feature testing macro __cpp_decltype_auto
to enable my changes. That points to a C++14 compliant compiler and is afaik the only c++14 feature I need. The macro is only standardized with C++20, but compilers support it before that. 

For now the stl functions all_of etc. and std::find are implemented. We now should have full support of forward and reverse iterators, with the same speedup as before. I added test cases for all of the functionality.

The other stl functions can be forwarded with the schema I made, it's just some doing from there on.

HattrickGenerator committed:
Guard the forwarded stl algorithms with feature testing macro, due to decltype auto usage

HattrickGenerator committed:
Fix implicit conversion in test to avoid warnings under windows build.

HattrickGenerator committed:
Change header names from .hh -> .hpp to be "GLOB"ed by CMakeLists.txt and appear in interface

HattrickGenerator commented:
@alalek Is it possible to reivew my changes again?
The build is successful now (except some warnings under Windows that seem to be caused by 3rd party etc.).
